src/D2L.CodeStyle.Analyzers/Helpers/ConstantAnalyzer.cs

```diff
@@ -0,0 +1,151 @@
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (InvocationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (ObjectCreationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private static void AnalyzeParameters(
+			SyntaxNodeAnalysisContext context,
+			ParameterListSyntax parameterList
+		) {
+			SeparatedSyntaxList<ParameterSyntax> parameters = parameterList.Parameters;
+			foreach( ParameterSyntax parameterSyntax in parameters ) {
+				// We need the symbol
+				IParameterSymbol parameter = context.SemanticModel.GetDeclaredSymbol( parameterSyntax );
+
+				// Parameter is somehow null, so do nothing
+				if( parameter == null ) {
+					continue;
+				}
+
+				// Parameter is not [Constant], so do nothing
+				if( !parameter.HasAttribute( AttributeName ) ) {
+					continue;
+				}
+
+				// Get the base type of the parameter
+				ITypeSymbol type = parameter.Type;
+
+				// Check that the parameter type is not a valid type,
+				// and that it is not a type parameter
+				if( type.SpecialType != SpecialType.None
+				 || type.Kind != SymbolKind.NamedType ) {
+					continue;
+				}
+
+				// This is marked as [Constant] and so you need to use a different type
+				context.ReportDiagnostic(
+					Diagnostic.Create(
+						descriptor: Diagnostics.InvalidConstantType,
+						location: parameterSyntax.GetLocation(),
+						messageArgs: type.TypeKind
+					)
+				);
+			}
+		}
+
+		private static void AnalyzeArguments(
+			SyntaxNodeAnalysisContext context,
+			ArgumentListSyntax argumentList ) {
+			// If there are no arguments, nothing needs to be done
+			if( argumentList == null ) {
+				return;
+			}
+
+			// Get arguments from the list and iterate through them
+			SeparatedSyntaxList<ArgumentSyntax> arguments = argumentList.Arguments;
+			foreach( var argument in arguments ) {
+				// Get the associated parameter
+				var parameter = argument.DetermineParameter(
+					context.SemanticModel,
+					allowParams: true
+				);
+
+				// Parameter is somehow null, so do nothing
+				if( parameter == null ) {
+					continue;
+				}
+
+				// Parameter is not [Constant], so do nothing
+				if( !parameter.HasAttribute( AttributeName ) ) {
+					continue;
+				}
+
+				// Get the base type of the argument
+				var type = context.SemanticModel.GetTypeInfo( argument.Expression ).Type;
+
+				// Check that the argument type is not a valid type
+				if( type?.SpecialType == SpecialType.None ) {
+					// This is marked as [Constant] and so you need to use a different type
+					context.ReportDiagnostic(
+						Diagnostic.Create(
+							descriptor: Diagnostics.InvalidConstantType,
+							location: argument.GetLocation(),
+							messageArgs: type.TypeKind
+						)
+					);
+
+					continue;
+				}

```

Comment:
```suggestion
```

This check doesn't really make sense on the argument does it? Made sense on the parameter to ensure the attribute is being used correctly, but for the argument the constant check below seems like it's what we want.

Reply:
I added a more descriptive comment.
Basically this is needed because when the parameter is generic (`SomeThing<T>([Constant] T obj) { }`) you can't tell if the type can be constant or not - that needs to be decided when the actual type is determined.

Reply:
But that's just confusing / uninteresting as a consumer of an API. It's already been marked as constant and there's nothing I can do about that. As a consumer the interesting thing is whether the value I'm passing is constant or not.

Checking the type the parameter is attached to and raising this diagnostic is just to keep people from applying to a parameter that can never be constant, and thus any usage of it would be an error.